### ISSUE TITLE
Aborting a git commit in a repo with pre-commit hooks from the git tab does not yield the same result as aborting from the command line

### ISSUE THREAD

**Issue by lorenzwalthert** [rstudio/rstudio#6471]
### System details

    RStudio Edition : Desktop
    RStudio Version : 1.2.5033
    OS Version      :  macOS 10.15.3
    R Version       :  3.6.1

### Steps to reproduce the problem

* clone example repo with minimal example: `git clone git@github.com:lorenzwalthert/rstudio-git-bug.git`
* Install the [pre-commit](https://pre-commit.com) framework with
  `remotes::install_github('lorenzwalthert/precommit')`
* initiate the hooks with `precommit::use_precommit()`.
* edit the two R files file1 and file2 and stage one.
* attempt to commit via the RStudio git tab, then abort immediately.
* The file that was not staged was stashed away during the pre-commit checks (by pre-commit) and should be restored afterwards according to pre-commit. This does not happen when committing via the RStudio git tab, but it happens when committing via the console.

### Describe the problem in detail

https://pre-commit.com is an [increasingly](https://star-history.t9t.io/#pre-commit/pre-commit) popular framework for managing git hooks, there are specific [R hooks](https://github.com/lorenzwalthert/precommit) too (disclosure: I am the maintainer). Upon committing, it stashes all non-staged changes to a cache. Aborting to commit with the RStudio git tab makes all changes that are not staged for commit lost (unless one is a git expert). This is not the case when aborting from the command line because pre-commit restores the stashes in this case from the cache.

Suspicion: Aborting the commit via the RStudio git tab is implemented with KILL and not with INT or similar (https://github.com/pre-commit/pre-commit/issues/1362#issuecomment-598434954) so pre-commit cannot do the cleanup. 

This behaviour makes it very dangerous to use pre-commit with RStudio and for that reason I think it should be fixed.

Reference: https://github.com/pre-commit/pre-commit/issues/1362

### Describe the behavior you expected

I'd expect the stash to be restored after aborting the commit, i.e. that aborting a git commit on the command line results in the same git status as aborting from the RStudio git tab.

- [x] I have read the guide to submitting good bug reports at https://github.com/rstudio/rstudio/wiki/Writing-Good-Bug-Reports . 
- [x] I have installed the latest version of RStudio and confirmed that the issue still persists.
- [x] If I am reporting a RStudio crash, I have included a diagnostics report. https://support.rstudio.com/hc/en-us/articles/200321257-Running-a-Diagnostics-Report
- [x] I have done my best to include a minimal, self-contained set of instructions for consistently reproducing the issue.

cc: @asottile

**Comment by ronblum** [comment]
@lorenzwalthert Thanks! I can reproduce this with the code/steps above. I'll bump this up to triage for review as we continue development of RStudio.

**Comment by lorenzwalthert** [comment]
Thanks for the heads up @ronblum. I believe this is possibly a quick fix.

**Comment by pat-s** [comment]
@ronblum Is there any rough ETA for this fix?

**Comment by kevinushey** [comment]
@lorenzwalthert, it looks like your intuition here is basically correct.

Just to log some investigation here -- we have a ConsoleProcess class that handles these sub-processes, launched e.g. here:

https://github.com/rstudio/rstudio/blob/1889a1bce5322cb541b88adfe56ecdefa1b43a20/src/cpp/session/modules/SessionGit.cpp#L430-L460

When a process is cancelled, we set an interrupt flag here:

https://github.com/rstudio/rstudio/blob/1889a1bce5322cb541b88adfe56ecdefa1b43a20/src/cpp/session/SessionConsoleProcess.cpp#L404-L407

Which is eventually checked and read within our `onContinue()` handler:

https://github.com/rstudio/rstudio/blob/1889a1bce5322cb541b88adfe56ecdefa1b43a20/src/cpp/session/SessionConsoleProcess.cpp#L422-L424

But having that return false actually implies terminating (as opposed to interrupting) the process:

https://github.com/rstudio/rstudio/blob/1889a1bce5322cb541b88adfe56ecdefa1b43a20/src/cpp/core/system/PosixChildProcess.cpp#L1029-L1039

It seems like this could be fixed by sending a real interrupt to the process, although we'd of course need to check whether the current behavior is intentional for certain contexts.

**Comment by lorenzwalthert** [comment]
@mikebessuille can you explain why this was  postponed? It seems from the upvotes this issue got that it is quite a popular request.

**Comment by mikebessuille** [comment]
@lorenzwalthert, absolutely.  This is an important issue, and it makes the cut for the near term.  In the past we would typically do very long releases (1y+).  We're moving to a model of smaller, more frequent releases; there are a lot of benefits of this.  Of course, this means we will deliver less in each release.  The 1.5 release as planned was growing out of control so that it would have taken us ~9 months to deliver.  So, we split that functionality across several milestones.   Our very short term focus is on replacing our use of an internal embedded browser library with Electron, as that will help resolve a multitude (over 50 at latest count) of UI related issues across the board.  So it's just a matter of needing to focus on that for 1.5, at the exclusion of almost everything else.

**Comment by llrs** [comment]
Hi, I'm planning on using git hooks via precommit but I want to be sure I can use them safely within RStudio. 
It's been a while since the last change on this issue (> 1 year). Is this fix coming or it needs more time?

**Comment by lorenzwalthert** [comment]
I don't know if there will be a fix soon. Although it has 35 upvotes, I feel like it has been postponed a few times, although the solution was already outlined by Kevin above and implementation effort seems not huge.

**Comment by lorenzwalthert** [comment]
@ronblum amazing. I already gave up (almost) 😀. Glad you are on it, the issue has 42 upvotes. pre-commit has to be installed using your package manager of choice. Common options on windows are pip(3) / pipx for windows. If you have `conda`, on your machine eg.g. because you may have used {reticulate}, you should be able to do `precommit::install_precommit()` to install and link the CLI executable.

**Comment by llrs** [comment]
I didn't manage to install correctly precommit (It works on one repository but not on others), but I did provide some examples and different configurations at https://github.com/lorenzwalthert/precommit/issues/605 Hope it helps to install it. Looking forward to when this works without so many issues on windows. Thanks!

**Comment by ronblum** [comment]
I had to install pre-commit with `pip install pre-commit` which I hadn't done. All set now.

@kevinushey I'm still seeing a failure. I'm aborting the process by closing the commit window while RStudio was committing, and the unstaged file didn't reappear in the Files Pane and Git Pane.

**Comment by kevinushey** [comment]
It works for me?

<img width="647" alt="Image" src="https://github.com/user-attachments/assets/a2df9cc2-f4de-4acd-8898-e83faec8285c" />

And then the files are staged / unstaged as they were before starting the git operation.

<img width="431" alt="Image" src="https://github.com/user-attachments/assets/c143a5b1-5378-470a-adde-2d8b03ac7ea9" />

Can you provide more details?

**Comment by ronblum** [comment]
I used the "Stop" button, clicking on it multiple times. Easier with a screen recording:

https://github.com/user-attachments/assets/3b018e0d-57af-443d-9ad1-01e1b2c85d43

**Comment by kevinushey** [comment]
Hmm -- it seems like the process did not handle the interrupt in time, so we were forced to terminate it.

**Comment by kevinushey** [comment]
As far as I can tell, we _are_ sending the interrupt to the running Git process (even on Windows), it's just getting ignored or dropped. @ronblum in your example, if you press the Stop button multiple times, then we terminate the process, so that explains what you're seeing. If you just press it once, you should see it basically ... do nothing.

I'd vote to just leave this alone since I imagine the majority of users doing something like use pre-commit would be on macOS or Linux?

**Comment by ronblum** [comment]
Sounds good. And "don't press the Stop button a thousand times" is good practice.

**Comment by lorenzwalthert** [comment]
Thanks guys for implementing this.
